openmdao/drivers/differential_evolution_driver.py

```diff
@@ -0,0 +1,593 @@
+                if extra > 0:
+                    for j in range(comm.size - extra):
+                        cases.append(cases[-1])
+
+                results = concurrent_eval(self.objfun, cases, comm,
+                                          allgather=True, model_mpi=self.model_mpi)
+
+                fitness[:] = np.inf
+                for result in results:
+                    returns, traceback = result
+
+                    if returns:
+                        val, success, ii = returns
+                        if success:
+                            fitness[ii] = val
+                            nfit += 1
+                    else:
+                        # Print the traceback if it fails
+                        print('A case failed:')
+                        print(traceback)
+            else:  # Serial
+                for ii in range(self.npop):
+                    fitness[ii], success, _ = self.objfun(population[ii], 0)
+                    nfit += 1
+
+            # Find best performing point in this generation.
+            min_fit = np.min(fitness)
+            min_index = np.argmin(fitness)
+            min_gen = population[min_index]
+
+            # Update overall best.
+            if min_fit < fopt:
+                fopt = min_fit
+                xopt = min_gen
+
+            if generation == max_gen:  # finished
+                break
+
+            # Selection: new generation members replace parents, if better (implied elitism)
+            if generation == 0:
+                parentPop = copy.deepcopy(population)
+                parentFitness = copy.deepcopy(fitness)
+            else:
+                for ii in range(self.npop):
+                    if fitness[ii] < parentFitness[ii]:  # if child is better, else parent unchanged
+                        parentPop[ii] = population[ii]
+                        parentFitness[ii] = fitness[ii]
+
+            # Evolve new generation.
+            population = np.zeros(np.shape(parentPop))
+            fitness = np.ones(self.npop) * np.inf
+
+            for ii in range(self.npop):
+                # randomly select 3 different population members other than the current choice
+                a, b, c = ii, ii, ii
+                while a == ii:
+                    a = rng.integers(0, self.npop)
+                while b == ii or b == a:
+                    b = rng.integers(0, self.npop)
+                while c == ii or c == a or c == b:
+                    c = rng.integers(0, self.npop)
+
+                # randomly select chromosome index for crossover
+                r = rng.integers(0, self.lchrom)
+
+                # crossover and mutation
+                population[ii] = parentPop[ii]  # start the same as parent
+                # clip mutant so that it cannot be outside the bounds
+                mutant = np.clip(parentPop[a] + F * (parentPop[b] - parentPop[c]), vlb, vub)
+                for j in range(self.lchrom):
```

Comment:
Nested for-loops are slow in python.  I think you could speed this up with some numpy vectorized calls.  Off the top of my head, something like this:

rr = rng.random(self.lchrom)
idx = np.where(rr < Pc)
population[ii][idx] = mutant[idx]
population[ii][r] = mutant[r]

vectorizing the 'ii' loop might also be possible, but it looks like a difficult task, and the resulting code might not be very readable.